File: python/packages/cip/src/new_odyssey_cip/records.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Literal

IsoTimestamp = str
ProductTier = Literal["pegasus", "pantheon", "phoenix"]
Environment = Literal["development", "test", "sandbox", "production"]
PolicyDomain = Literal["platform", "security", "expense", "recruitment", "onboarding"]
PolicyOperator = Literal[
    "eq",
    "neq",
    "in",
    "contains",
    "exists",
    "regex",
    "gt",
    "gte",
    "lt",
    "lte",
]
PolicyRuleAction = Literal["allow", "flag", "block", "escalate"]
Severity = Literal["info", "warn", "high", "critical"]
ReleaseState = Literal["draft", "released", "deprecated"]
DeploymentStatus = Literal[
    "provisioning",
    "active",
    "paused",
    "draining",
    "failed",
    "retired",
]
RunSessionStatus = Literal[
    "queued",
    "running",
    "waiting-human",
    "completed",
    "failed",
]
ComplianceRegime = Literal["eu-ai-act"]
ComplianceRiskTier = Literal["minimal", "limited", "high-risk", "prohibited", "unclassified"]
ApprovalRequestStatus = Literal[
    "pending",
    "approved",
    "rejected",
    "expired",
    "cancelled",
]
RunEventType = Literal[
    "run_started",
    "tool_called",
    "tool_completed",
    "handoff_started",
    "handoff_completed",
    "guardrail_triggered",
    "policy_decided",
    "approval_requested",
    "approval_resolved",
    "disclosure_presented",
    "disclosure_acknowledged",
    "human_review_completed",
    "output_overridden",
    "stop_invoked",
    "run_completed",
    "run_failed",
]
ActorVerification = Literal[
    "system",
    "authenticated-sdk",
    "authenticated-operator",
    "asserted",
]
ComplianceArtifactKind = Literal[
    "technical_documentation",
    "fundamental_rights_impact_assessment",
    "conformity_assessment",
    "eu_declaration_of_conformity",
    "eu_database_registration",
    "post_market_monitoring_plan",
    "serious_incident_record",
]
ComplianceArtifactStatus = Literal["draft", "approved", "filed", "not_applicable", "expired"]
DisclosureSurface = Literal["banner", "first_message", "banner_and_first_message"]
GuardrailKey = Literal[
    "tenant_boundary",
    "pii_boundary",
    "least_privilege",
    "manual_review_required",
    "data_residency",
]


@dataclass(slots=True)
class BaseRecord:
    id: str
    created_at: IsoTimestamp
    updated_at: IsoTimestamp
    revision: int


@dataclass(slots=True)
class TraceCorrelation:
    provider: Literal["openai", "custom"]
    trace_id: str | None = None
    span_id: str | None = None
    conversation_id: str | None = None
    response_id: str | None = None


@dataclass(slots=True)
class HighRiskBasis:
    annex: Literal["annex-i", "annex-iii"]
    category: str
    rationale: str


@dataclass(slots=True)
class ComplianceTransparency:
    required: bool
    notice_text: str
    placement: Literal["banner-and-first-message"]
    requires_acknowledgement: bool


@dataclass(slots=True)
class ComplianceOversight:
    required: bool
    require_approval_before_completion: bool
    minimum_human_reviewers: int
    stop_mechanism_required: bool


@dataclass(slots=True)
class ComplianceLogging:
    require_verified_actors: bool
    retention_days: int


@dataclass(slots=True)
class TenantRecord(BaseRecord):
    slug: str
    display_name: str
    product_tier: ProductTier
    platforms: list[str]
    regions: list[str]
    status: Literal["active", "suspended", "retired"]


@dataclass(slots=True)
class ConnectorDefinition(BaseRecord):
    key: str
    version: str
    platform: str
    display_name: str
    runtime: Literal["mcp", "native", "http"]
    auth_strategy: Literal["oauth2", "api-key", "service-account", "custom"]
    source: Literal["first-party", "partner", "community"]
    capabilities: list[str]
    status: Literal["draft", "active", "deprecated"]
    driver_key: str | None = None
    driver_config: dict[str, Any] | None = None
    metadata: dict[str, Any] | None = None


@dataclass(slots=True)
class CredentialBinding(BaseRecord):
    tenant_id: str
    name: str
    provider: str
    secret_ref: str
    scopes: list[str]
    status: Literal["active", "rotated", "revoked"]
    secret_backend_key: str | None = None
    expires_at: IsoTimestamp | None = None
    rotated_at: IsoTimestamp | None = None


@dataclass(slots=True)
class ConnectorBinding(BaseRecord):
    tenant_id: str
    connector_definition_id: str
    credential_binding_id: str
    environment: Environment
    alias: str
    endpoint: str
    config: dict[str, Any]
    status: Literal["active", "disabled"]


@dataclass(slots=True)
class PolicyCondition:
    path: str
    operator: PolicyOperator
    value: Any | None = None


@dataclass(slots=True)
class PolicyClause:
    id: str
    name: str
    match: Literal["all", "any"]
    conditions: list[PolicyCondition]


@dataclass(slots=True)
class PolicyRule:
    id: str
    name: str
    severity: Severity
    action: PolicyRuleAction
    description: str | None = None
    expression: str | None = None
    clauses: list[PolicyClause] = field(default_factory=list)


@dataclass(slots=True)
class PolicyPack(BaseRecord):
    key: str
    name: str
    domain: PolicyDomain
    version: str
    ownership: Literal["shared", "tenant"]
    rules: list[PolicyRule]
    guardrail_refs: list[str]
    status: Literal["draft", "active", "retired"]
    tenant_id: str | None = None


@dataclass(slots=True)
class RuntimeProfile:
    provider: Literal["openai-agents-sdk", "anthropic", "custom"]
    model_profile: Literal["default", "reasoning", "fast"]
    adapter_version: str | None = None


@dataclass(slots=True)
class DependencyVersionReference:
    id: str
    key: str
    version: str


@dataclass(slots=True)
class BlueprintDependencySnapshot:
    policy_packs: list[DependencyVersionReference]
    guardrails: list[DependencyVersionReference]
    connector_manifests: list[DependencyVersionReference]
    runtime_adapter_version: str


@dataclass(slots=True)
class AgentBlueprint(BaseRecord):
    key: str
    version: str
    release_state: ReleaseState
    dependency_snapshot: BlueprintDependencySnapshot
    name: str
    product_tier: ProductTier
    domain: PolicyDomain
    description: str
    runtime: RuntimeProfile
    connector_definition_ids: list[str]
    policy_pack_ids: list[str]
    guardrail_definition_ids: list[str]
    handoff_targets: list[str]
    status: Literal["draft", "active", "deprecated"]
    supersedes_blueprint_id: str | None = None
    structured_output: str | None = None


@dataclass(slots=True)
class DeploymentRecord(BaseRecord):
    tenant_id: str
    agent_blueprint_id: str
    agent_blueprint_version: str
    environment: Environment
    connector_binding_ids: list[str]
    policy_pack_ids: list[str]
    status: DeploymentStatus
    tags: list[str]
    last_transition_at: IsoTimestamp
    deployed_at: IsoTimestamp | None = None


@dataclass(slots=True)
class ComplianceProfile(BaseRecord):
    tenant_id: str
    deployment_id: str
    regime: ComplianceRegime
    serves_eu_users: bool
    intended_purpose: str
    risk_tier: ComplianceRiskTier
    transparency: ComplianceTransparency
    oversight: ComplianceOversight
    logging: ComplianceLogging
    high_risk_basis: HighRiskBasis | None = None


@dataclass(slots=True)
class ComplianceArtifact(BaseRecord):
    tenant_id: str
    deployment_id: str
    kind: ComplianceArtifactKind
    status: ComplianceArtifactStatus
    owner: str
    summary: str
    external_ref: str | None = None
    due_at: IsoTimestamp | None = None
    completed_at: IsoTimestamp | None = None


@dataclass(slots=True)
class RunSession(BaseRecord):
    tenant_id: str
    deployment_id: str
    correlation_id: str
    status: RunSessionStatus
    started_at: IsoTimestamp
    input_summary: str
    completed_at: IsoTimestamp | None = None
    output_summary: str | None = None
    current_approval_request_id: str | None = None
    trace_correlation: TraceCorrelation | None = None
    compliance_profile_snapshot: ComplianceProfile | None = None


@dataclass(slots=True)
class AuditActor:
    type: Literal["agent", "human", "system"]
    id: str


@dataclass(slots=True)
class AuditEvent:
    id: str
    tenant_id: str
    category: Literal[
        "tenant",
        "connector",
        "policy",
        "deployment",
        "session",
        "security",
        "approval",
        "runtime",
    ]
    action: str
    severity: Literal["info", "warn", "error", "critical"]
    occurred_at: IsoTimestamp
    actor: AuditActor
    actor_verification: ActorVerification
    payload: dict[str, Any]
    asserted_actor: AuditActor | None = None
    deployment_id: str | None = None
    session_id: str | None = None


@dataclass(slots=True)
class GuardrailDefinition(BaseRecord):
    key: str
    version: str
    name: str
    configuration: dict[str, Any]
    status: Literal["draft", "active", "retired"]
    description: str | None = None


@dataclass(slots=True)
class ApprovalRequest(BaseRecord):
    tenant_id: str
    deployment_id: str
    session_id: str
    checkpoint_id: str
    reason: str
    requested_by: AuditActor
    status: ApprovalRequestStatus
    expires_at: IsoTimestamp | None = None
    resolved_at: IsoTimestamp | None = None
    resolution_comment: str | None = None
    guardrail_definition_id: str | None = None
    policy_pack_id: str | None = None


@dataclass(slots=True)
class RunEvent(BaseRecord):
    tenant_id: str
    deployment_id: str
    session_id: str
    type: RunEventType
    sequence: int
    occurred_at: IsoTimestamp
    actor: AuditActor
    actor_verification: ActorVerification
    payload: dict[str, Any]
    asserted_actor: AuditActor | None = None
    trace_correlation: TraceCorrelation | None = None


@dataclass(slots=True)
class DisclosureRecord(BaseRecord):
    tenant_id: str
    deployment_id: str
    session_id: str
    disclosure_version: str
    surface: DisclosureSurface
    presented_at: IsoTimestamp
    acknowledged_at: IsoTimestamp | None = None


@dataclass(slots=True)
class HumanReviewRecord(BaseRecord):
    tenant_id: str
    deployment_id: str
    session_id: str
    reviewer: AuditActor
    decision: Literal["approved", "rejected"]
    reviewed_at: IsoTimestamp
    comment: str | None = None


@dataclass(slots=True)
class EvidenceBundle(BaseRecord):
    tenant_id: str
    deployment_id: str
    session_id: str
    agent_blueprint_id: str
    agent_blueprint_version: str
    policy_pack_versions: list[DependencyVersionReference]
    guardrail_versions: list[DependencyVersionReference]
    summary: str
    run_event_ids: list[str]
    audit_event_ids: list[str]
    disclosure_record_ids: list[str]
    human_review_ids: list[str]
    compliance_artifact_ids: list[str]
    generated_at: IsoTimestamp
    compliance_profile: ComplianceProfile | None = None


@dataclass(slots=True)
class ConnectorRateBucket(BaseRecord):
    provider: str
    external_system_tenant: str
    environment: Environment
    api_family: str
    max_requests_per_second: int
    available_tokens: float
    last_refill_at: IsoTimestamp
    queue_depth: int
    status: Literal["active", "disabled"]

File: typescript/packages/cip/src/domain/records.ts
export type IsoTimestamp = string;

export type ProductTier = "pegasus" | "pantheon" | "phoenix";
export type Environment = "development" | "test" | "sandbox" | "production";
export type PolicyDomain =
  | "platform"
  | "security"
  | "expense"
  | "recruitment"
  | "onboarding";
export type PolicyOperator =
  | "eq"
  | "neq"
  | "in"
  | "contains"
  | "exists"
  | "regex"
  | "gt"
  | "gte"
  | "lt"
  | "lte";
export type PolicyRuleAction = "allow" | "flag" | "block" | "escalate";
export type Severity = "info" | "warn" | "high" | "critical";
export type ReleaseState = "draft" | "released" | "deprecated";
export type DeploymentStatus =
  | "provisioning"
  | "active"
  | "paused"
  | "draining"
  | "failed"
  | "retired";
export type RunSessionStatus =
  | "queued"
  | "running"
  | "waiting-human"
  | "completed"
  | "failed";
export type ComplianceRegime = "eu-ai-act";
export type ComplianceRiskTier =
  | "minimal"
  | "limited"
  | "high-risk"
  | "prohibited"
  | "unclassified";
export type ApprovalRequestStatus =
  | "pending"
  | "approved"
  | "rejected"
  | "expired"
  | "cancelled";
export type RunEventType =
  | "run_started"
  | "tool_called"
  | "tool_completed"
  | "handoff_started"
  | "handoff_completed"
  | "guardrail_triggered"
  | "policy_decided"
  | "approval_requested"
  | "approval_resolved"
  | "disclosure_presented"
  | "disclosure_acknowledged"
  | "human_review_completed"
  | "output_overridden"
  | "stop_invoked"
  | "run_completed"
  | "run_failed";
export type ActorVerification =
  | "system"
  | "authenticated-sdk"
  | "authenticated-operator"
  | "asserted";
export type ComplianceArtifactKind =
  | "technical_documentation"
  | "fundamental_rights_impact_assessment"
  | "conformity_assessment"
  | "eu_declaration_of_conformity"
  | "eu_database_registration"
  | "post_market_monitoring_plan"
  | "serious_incident_record";
export type ComplianceArtifactStatus =
  | "draft"
  | "approved"
  | "filed"
  | "not_applicable"
  | "expired";
export type DisclosureSurface =
  | "banner"
  | "first_message"
  | "banner_and_first_message";
export type GuardrailKey =
  | "tenant_boundary"
  | "pii_boundary"
  | "least_privilege"
  | "manual_review_required"
  | "data_residency"
  | string;

export interface BaseRecord {
  id: string;
  createdAt: IsoTimestamp;
  updatedAt: IsoTimestamp;
  revision: number;
}

export interface TraceCorrelation {
  provider: "openai" | "custom";
  traceId?: string;
  spanId?: string;
  conversationId?: string;
  responseId?: string;
}

export interface HighRiskBasis {
  annex: "annex-i" | "annex-iii";
  category: string;
  rationale: string;
}

export interface ComplianceTransparency {
  required: boolean;
  noticeText: string;
  placement: "banner-and-first-message";
  requiresAcknowledgement: boolean;
}

export interface ComplianceOversight {
  required: boolean;
  requireApprovalBeforeCompletion: boolean;
  minimumHumanReviewers: number;
  stopMechanismRequired: boolean;
}

export interface ComplianceLogging {
  requireVerifiedActors: boolean;
  retentionDays: number;
}

export interface TenantRecord extends BaseRecord {
  slug: string;
  displayName: string;
  productTier: ProductTier;
  platforms: string[];
  regions: string[];
  status: "active" | "suspended" | "retired";
}

export interface ConnectorDefinition extends BaseRecord {
  key: string;
  version: string;
  platform: string;
  displayName: string;
  driverKey?: string;
  driverConfig?: Record<string, unknown>;
  runtime: "mcp" | "native" | "http";
  authStrategy: "oauth2" | "api-key" | "service-account" | "custom";
  source: "first-party" | "partner" | "community";
  capabilities: string[];
  status: "draft" | "active" | "deprecated";
  metadata?: Record<string, unknown>;
}

export interface CredentialBinding extends BaseRecord {
  tenantId: string;
  name: string;
  provider: string;
  secretBackendKey?: string;
  secretRef: string;
  scopes: string[];
  expiresAt?: IsoTimestamp;
  rotatedAt?: IsoTimestamp;
  status: "active" | "rotated" | "revoked";
}

export interface ConnectorBinding extends BaseRecord {
  tenantId: string;
  connectorDefinitionId: string;
  credentialBindingId: string;
  environment: Environment;
  alias: string;
  endpoint: string;
  config: Record<string, unknown>;
  status: "active" | "disabled";
}

export interface PolicyCondition {
  path: string;
  operator: PolicyOperator;
  value?: unknown;
}

export interface PolicyClause {
  id: string;
  name: string;
  match: "all" | "any";
  conditions: PolicyCondition[];
}

export interface PolicyRule {
  id: string;
  name: string;
  description?: string;
  expression?: string;
  clauses?: PolicyClause[];
  severity: Severity;
  action: PolicyRuleAction;
}

export interface PolicyPack extends BaseRecord {
  key: string;
  name: string;
  domain: PolicyDomain;
  version: string;
  ownership: "shared" | "tenant";
  tenantId?: string;
  rules: PolicyRule[];
  guardrailRefs: string[];
  status: "draft" | "active" | "retired";
}

export interface RuntimeProfile {
  provider: "openai-agents-sdk" | "anthropic" | "custom";
  modelProfile: "default" | "reasoning" | "fast";
  adapterVersion?: string;
}

export interface DependencyVersionReference {
  id: string;
  key: string;
  version: string;
}

export interface BlueprintDependencySnapshot {
  policyPacks: DependencyVersionReference[];
  guardrails: DependencyVersionReference[];
  connectorManifests: DependencyVersionReference[];
  runtimeAdapterVersion: string;
}

export interface AgentBlueprint extends BaseRecord {
  key: string;
  version: string;
  releaseState: ReleaseState;
  supersedesBlueprintId?: string;
  dependencySnapshot: BlueprintDependencySnapshot;
  name: string;
  productTier: ProductTier;
  domain: PolicyDomain;
  description: string;
  runtime: RuntimeProfile;
  connectorDefinitionIds: string[];
  policyPackIds: string[];
  guardrailDefinitionIds: string[];
  handoffTargets: string[];
  structuredOutput?: string;
  status: "draft" | "active" | "deprecated";
}

export interface DeploymentRecord extends BaseRecord {
  tenantId: string;
  agentBlueprintId: string;
  agentBlueprintVersion: string;
  environment: Environment;
  connectorBindingIds: string[];
  policyPackIds: string[];
  status: DeploymentStatus;
  deployedAt?: IsoTimestamp;
  lastTransitionAt: IsoTimestamp;
  tags: string[];
}

export interface ComplianceProfile extends BaseRecord {
  tenantId: string;
  deploymentId: string;
  regime: ComplianceRegime;
  servesEuUsers: boolean;
  intendedPurpose: string;
  riskTier: ComplianceRiskTier;
  highRiskBasis?: HighRiskBasis;
  transparency: ComplianceTransparency;
  oversight: ComplianceOversight;
  logging: ComplianceLogging;
}

export interface ComplianceArtifact extends BaseRecord {
  tenantId: string;
  deploymentId: string;
  kind: ComplianceArtifactKind;
  status: ComplianceArtifactStatus;
  owner: string;
  summary: string;
  externalRef?: string;
  dueAt?: IsoTimestamp;
  completedAt?: IsoTimestamp;
}

export interface RunSession extends BaseRecord {
  tenantId: string;
  deploymentId: string;
  correlationId: string;
  status: RunSessionStatus;
  startedAt: IsoTimestamp;
  completedAt?: IsoTimestamp;
  inputSummary: string;
  outputSummary?: string;
  currentApprovalRequestId?: string;
  traceCorrelation?: TraceCorrelation;
  complianceProfileSnapshot?: ComplianceProfile | null;
}

export interface AuditActor {
  type: "agent" | "human" | "system";
  id: string;
}

export interface AuditEvent {
  id: string;
  tenantId: string;
  deploymentId?: string;
  sessionId?: string;
  category:
    | "tenant"
    | "connector"
    | "policy"
    | "deployment"
    | "session"
    | "security"
    | "approval"
    | "runtime";
  action: string;
  severity: "info" | "warn" | "error" | "critical";
  occurredAt: IsoTimestamp;
  actor: AuditActor;
  assertedActor?: AuditActor;
  actorVerification: ActorVerification;
  payload: Record<string, unknown>;
}

export interface GuardrailDefinition extends BaseRecord {
  key: GuardrailKey;
  version: string;
  name: string;
  description?: string;
  configuration: Record<string, unknown>;
  status: "draft" | "active" | "retired";
}

export interface ApprovalRequest extends BaseRecord {
  tenantId: string;
  deploymentId: string;
  sessionId: string;
  checkpointId: string;
  reason: string;
  requestedBy: AuditActor;
  status: ApprovalRequestStatus;
  expiresAt?: IsoTimestamp;
  resolvedAt?: IsoTimestamp;
  resolutionComment?: string;
  guardrailDefinitionId?: string;
  policyPackId?: string;
}

export interface RunEvent extends BaseRecord {
  tenantId: string;
  deploymentId: string;
  sessionId: string;
  type: RunEventType;
  sequence: number;
  occurredAt: IsoTimestamp;
  actor: AuditActor;
  assertedActor?: AuditActor;
  actorVerification: ActorVerification;
  payload: Record<string, unknown>;
  traceCorrelation?: TraceCorrelation;
}

export interface DisclosureRecord extends BaseRecord {
  tenantId: string;
  deploymentId: string;
  sessionId: string;
  disclosureVersion: string;
  surface: DisclosureSurface;
  presentedAt: IsoTimestamp;
  acknowledgedAt?: IsoTimestamp;
}

export interface HumanReviewRecord extends BaseRecord {
  tenantId: string;
  deploymentId: string;
  sessionId: string;
  reviewer: AuditActor;
  decision: "approved" | "rejected";
  comment?: string;
  reviewedAt: IsoTimestamp;
}

export interface EvidenceBundle extends BaseRecord {
  tenantId: string;
  deploymentId: string;
  sessionId: string;
  agentBlueprintId: string;
  agentBlueprintVersion: string;
  policyPackVersions: DependencyVersionReference[];
  guardrailVersions: DependencyVersionReference[];
  summary: string;
  runEventIds: string[];
  auditEventIds: string[];
  complianceProfile?: ComplianceProfile | null;
  disclosureRecordIds: string[];
  humanReviewIds: string[];
  complianceArtifactIds: string[];
  generatedAt: IsoTimestamp;
}

export interface ConnectorRateBucket extends BaseRecord {
  provider: string;
  externalSystemTenant: string;
  environment: Environment;
  apiFamily: string;
  maxRequestsPerSecond: number;
  availableTokens: number;
  lastRefillAt: IsoTimestamp;
  queueDepth: number;
  status: "active" | "disabled";
}
